### ISSUE TITLE
App updates in release mode but not in debug mode

### ISSUE THREAD

**Issue by Areiser** [flutter/flutter#14527]
## Steps to Reproduce

After updating my flutter install, migrating to the new gradle plugin and updating my dependencies (share and admob plugin), my app only runs the current version if I use flutter run --release. When running flutter run, the old app is displayed even though I add new buttons etc.

In the log, I get an error that does not really tell me anything. Logs for both cases are attached. I'm at a loss to what I can do here. For migrating I followed this guide:  https://github.com/flutter/flutter/wiki/Updating-Flutter-projects-to-Gradle-4.1-and-Android-Studio-Gradle-plugin-3.0.1

## Logs
[Flutter_release.txt](https://github.com/flutter/flutter/files/1704347/flutter_release.txt)
[Flutter_debug.txt](https://github.com/flutter/flutter/files/1704348/Flutter_debug.txt)
[Flutter_analyze.txt](https://github.com/flutter/flutter/files/1704350/Flutter_analyze.txt)

## Flutter Doctor

[andreas@andreas-desktop Idiotentest]$ flutter doctor -v
[  +17 ms] [/run/media/andreas/Andreas/Programmieren/flutter/] git rev-parse --abbrev-ref --symbolic @{u}
[  +19 ms] Exit code 0 from: git rev-parse --abbrev-ref --symbolic @{u}
[        ] origin/alpha
[        ] [/run/media/andreas/Andreas/Programmieren/flutter/] git rev-parse --abbrev-ref HEAD
[   +2 ms] Exit code 0 from: git rev-parse --abbrev-ref HEAD
[        ] alpha
[        ] [/run/media/andreas/Andreas/Programmieren/flutter/] git ls-remote --get-url origin
[   +4 ms] Exit code 0 from: git ls-remote --get-url origin
[        ] https://github.com/flutter/flutter.git
[        ] [/run/media/andreas/Andreas/Programmieren/flutter/] git log -n 1 --pretty=format:%H
[   +5 ms] Exit code 0 from: git log -n 1 --pretty=format:%H
[        ] 2e449f06f0a3be076e336ad6b30b0e9ec99dbdfe
[        ] [/run/media/andreas/Andreas/Programmieren/flutter/] git log -n 1 --pretty=format:%ar
[   +3 ms] Exit code 0 from: git log -n 1 --pretty=format:%ar
[        ] 9 days ago
[        ] [/run/media/andreas/Andreas/Programmieren/flutter/] git describe --match v*.*.* --first-parent --long --tags
[   +3 ms] Exit code 0 from: git describe --match v*.*.* --first-parent --long --tags
[        ] v0.0.21-0-g2e449f06f
[ +152 ms] [✓] Flutter (on Linux, locale en_US.UTF-8, channel alpha)
[        ]     • Flutter version 0.0.21 at /run/media/andreas/Andreas/Programmieren/flutter
[        ]     • Framework revision 2e449f06f0 (9 days ago), 2018-01-29 14:26:51 -0800
[        ]     • Engine revision 6921873c71
[        ]     • Tools Dart version 2.0.0-dev.16.0
[        ]     • Engine Dart version 2.0.0-edge.da1f52592ef73fe3afa485385cb995b9aec0181a
[  +80 ms] java -version
[  +35 ms] [✓] Android toolchain - develop for Android devices (Android SDK 27.0.3)
[        ]     • Android SDK at /home/andreas/Android/Sdk
[        ]     • Android NDK at /home/andreas/Android/Sdk/ndk-bundle
[        ]     • Platform android-27, build-tools 27.0.3
[        ]     • Java binary at: /opt/android-studio/jre/bin/java
[        ]     • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-915-b01)
[   +1 ms] [✓] Android Studio (version 3.0)
[        ]     • Android Studio at /opt/android-studio
[        ]     • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-915-b01)
[  +54 ms] [✓] IntelliJ IDEA Community Edition (version 2017.3)
[        ]     • Flutter plugin version 21.2.3
[        ]     • Dart plugin version 173.4548.30
[   +3 ms] /home/andreas/Android/Sdk/platform-tools/adb devices -l
[   +3 ms] Exit code 0 from: /home/andreas/Android/Sdk/platform-tools/adb devices -l
[        ] List of devices attached
           FUH7N17220000709       device usb:3-5 product:VNS-L31 model:HUAWEI_VNS_L31 device:HWVNS-H
[  +10 ms] /home/andreas/Android/Sdk/platform-tools/adb -s FUH7N17220000709 shell getprop
[  +58 ms] ro.hardware = hi6250
[        ] ro.build.characteristics = default
[   +1 ms] [✓] Connected devices
[        ]     • HUAWEI VNS L31 • FUH7N17220000709 • android-arm • Android 7.0 (API 24)
[   +9 ms] "flutter doctor" took 370ms.
[ +175 ms] ensureAnalyticsSent: 173ms
[   +3 ms] exiting with code 0

**Comment by najeira** [comment]
did you try `flutter clean`?

**Comment by Areiser** [comment]
Oh wow. I didn't know that exists, thanks! I probably should have tried removing the build directory ;)

**Comment by Hixie** [comment]
You shouldn't need to call `flutter clean`.

Can you reproduce this? If we can figure out what causes us to get confused, we should fix it.

**Comment by Areiser** [comment]
Sorry for the delayed answer, but I tried this again today. Luckily I have it all versioned in git. It can reproduce it every time when jumping between commits.

1. Checkout old version of my app and run flutter clean (one local plugin in pubspec points to a local admob plugin):
`      path: ../plugins/packages/firebase_admob/
`

1. Build it with the current flutter version, this creates a bunch of errors probably related ot the older plugin I am using
-> realize I have to update gradle and the plugins.

1. Next I did multiple steps at once: Change the plugin from local to dartpub and use a newer version, update gradle, create a signingconfig with my keystore and add one button in my app

1. The problem initially mentioned in this issue occurs.

I have an idea why this could be happening though: When I migrated to the new gradle I noticed that my app had `com.yourcompany.projectname` instead of my actual package name (at this point the yourcompany package didn't even exist), so I changed it to my actual package. Could be related to this since the app does not update when running in non-release mode? 

[Diff that creates the problem](https://github.com/flutter/flutter/files/1750650/diff_gradleupdate.txt)

I hope this helps.
